client/sharing_url.go

```diff
@@ -0,0 +1,106 @@
+/***************************************************************
+ *
+ * Copyright (C) 2023, University of Nebraska-Lincoln
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you
+ * may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ***************************************************************/
+
+package client
+
+import (
+	"net/url"
+	"strings"
+
+	"github.com/pelicanplatform/pelican/config"
+	"github.com/pelicanplatform/pelican/param"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/viper"
+)
+
+func getDirectorFromUrl(objectUrl *url.URL) (string, error) {
+	configDirectorUrl := param.Federation_DirectorUrl.GetString()
+	var directorUrl string
+	if objectUrl.Scheme == "pelican" {
+		if objectUrl.Host == "" {
+			if configDirectorUrl == "" {
+				return "", errors.New("Must specify (or configure) the federation hostname with the pelican://-style URLs")
+			}
+			directorUrl = configDirectorUrl
+		} else {
+			discoveryUrl := url.URL{
+				Scheme: "https",
+				Host:   objectUrl.Host,
+			}
+			viper.Set("Federation.DirectorUrl", "")
+			viper.Set("Federation.DiscoveryUrl", discoveryUrl.String())
+			if err := config.DiscoverFederation(); err != nil {
+				return "", errors.Wrapf(err, "Failed to discover location of the director for the federation %s", objectUrl.Host)
+			}
+			if directorUrl = param.Federation_DirectorUrl.GetString(); directorUrl == "" {
+				return "", errors.Errorf("Director for the federation %s not discovered", objectUrl.Host)
+			}
+		}
+	} else if objectUrl.Scheme == "osdf" && configDirectorUrl == "" {
```

Comment:
Small bug that I happened to hit here because I forgot to clear my config. If we have the federation defined in the config and we are using `osdf://` as our scheme, we skip this block and hit line 74, telling us the scheme `osdf://` is invalid -- even if the federation information from the config is identical to the OSDF configuration. The mere fact it's defined in the configuration causes this to fail.

Reply:
Resolved in the latest push.